pandas/core/dtypes/astype.py

```diff
@@ -144,11 +144,18 @@ def astype_nansafe(
         if is_datetime64_dtype(dtype):
             from pandas import to_datetime
 
+            datetime_values = to_datetime(arr.ravel()).values.reshape(arr.shape)
+
+            if dtype == "<M8[D]":
```

Comment:
i am not in love with this special case. can you explain why this is the solution?

Reply:
Hi, sorry for late reply. The reason I place the special case for dtype == "<M8[D]", there is integer overflow in daily Series.astype('datetime64[D]') type. The integer overflow will happen when date before 1677-09-21.

Reply:
I only notice this today. If I test with date before `9/22/1677`, for example `9/21/1677`.  It will show this error
```
Traceback (most recent call last):
  File "d:\kcw\pandas\pandas\core\arrays\datetimes.py", line 2214, in objects_to_datetime64ns
    values, tz_parsed = conversion.datetime_to_datetime64(data.ravel("K"))
  File "pandas\_libs\tslibs\conversion.pyx", line 360, in pandas._libs.tslibs.conversion.datetime_to_datetime64
TypeError: Unrecognized value type: <class 'str'>
  File "d:\kcw\pandas\pandas\core\tools\datetimes.py", line 1077, in to_datetime
    result = convert_listlike(arg, format)
  File "d:\kcw\pandas\pandas\core\tools\datetimes.py", line 403, in _convert_listlike_datetimes
    result, tz_parsed = objects_to_datetime64ns(
  File "d:\kcw\pandas\pandas\core\arrays\datetimes.py", line 2220, in objects_to_datetime64ns
    raise err
  File "d:\kcw\pandas\pandas\core\arrays\datetimes.py", line 2202, in objects_to_datetime64ns
    result, tz_parsed = tslib.array_to_datetime(
  File "pandas\_libs\tslib.pyx", line 381, in pandas._libs.tslib.array_to_datetime
  File "pandas\_libs\tslib.pyx", line 611, in pandas._libs.tslib.array_to_datetime
  File "pandas\_libs\tslib.pyx", line 607, in pandas._libs.tslib.array_to_datetime
  File "pandas\_libs\tslib.pyx", line 562, in pandas._libs.tslib.array_to_datetime
  File "pandas\_libs\tslibs\conversion.pyx", line 517, in pandas._libs.tslibs.conversion.convert_datetime_to_tsobject
  File "pandas\_libs\tslibs\np_datetime.pyx", line 120, in pandas._libs.tslibs.np_datetime.check_dts_bounds
pandas._libs.tslibs.np_datetime.OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 1677-09-21 00:00:00
```

Reply:
> It will show this error

until we support non-nano dt64, this error is the _correct_ behavior

Reply:
@jbrockmendel is this where you want to locate? e.g. make this a function?

Reply:
i think dtypes.cast._ensure_nanosecond_dtype might be what we want here?

Reply:
I have look into dtypes.cast._ensure_nanosecond_dtype and inside code. Have idea how to start this. Will do this tomorrow.